src/debug_info.rs

```diff
@@ -1,30 +1,1060 @@
+        debug_loc: DILocation<'ctx>,
+        instruction: InstructionValue<'ctx>,
+    ) -> InstructionValue<'ctx> {
+        use llvm_sys::debuginfo::LLVMDIBuilderInsertDbgValueBefore;
+
+        let value_ref = unsafe {
+            LLVMDIBuilderInsertDbgValueBefore(
+                self.builder,
+                value.as_value_ref(),
+                var_info.metadata_ref,
+                expr.unwrap_or(self.create_expression(vec![])).metadata_ref,
+                debug_loc.metadata_ref,
+                instruction.as_value_ref(),
+            )
+        };
+        InstructionValue::new(value_ref)
+    }
+
+    /// Construct any deferred debug info descriptors. May generate invalid metadata if debug info
+    /// is incomplete.  Module/function verification can then fail.
+    ///
+    /// Call before any kind of code generation (including verification). Can be called more than
+    /// one time.
+    pub fn finalize(&self) {
+        use llvm_sys::debuginfo::LLVMDIBuilderFinalize;
+        unsafe { LLVMDIBuilderFinalize(self.builder) };
     }
 }
 
-impl Drop for DebugInfoBuilder {
+impl<'ctx> Drop for DebugInfoBuilder<'ctx> {
     fn drop(&mut self) {
-        // FIXME: Docs say DIB must be finalize before calling this
+        use llvm_sys::debuginfo::LLVMDisposeDIBuilder;
+
+        self.finalize();
+        unsafe { LLVMDisposeDIBuilder(self.builder) }
+    }
+}
+
+/// Source file scope for debug info
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub struct DIFile<'ctx> {
+    pub(crate) metadata_ref: LLVMMetadataRef,
+    _marker: PhantomData<&'ctx Context>,
+}
+
+impl<'ctx> AsDIScope<'ctx> for DIFile<'ctx> {
+    fn as_debug_info_scope(self) -> DIScope<'ctx> {
+        DIScope {
+            metadata_ref: self.metadata_ref,
+            _marker: PhantomData,
+        }
+    }
+}
+
+/// Compilation unit scope for debug info
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub struct DICompileUnit<'ctx> {
+    file: DIFile<'ctx>,
+    pub(crate) metadata_ref: LLVMMetadataRef,
+    _marker: PhantomData<&'ctx Context>,
+}
+
+impl<'ctx> DICompileUnit<'ctx> {
+    pub fn get_file(&self) -> DIFile<'ctx> {
+        self.file
+    }
+}
+
+impl<'ctx> AsDIScope<'ctx> for DICompileUnit<'ctx> {
+    fn as_debug_info_scope(self) -> DIScope<'ctx> {
+        DIScope {
+            metadata_ref: self.metadata_ref,
+            _marker: PhantomData,
+        }
+    }
+}
+
+/// Function body scope for debug info
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub struct DISubProgram<'ctx> {
+    pub(crate) metadata_ref: LLVMMetadataRef,
+    pub(crate) _marker: PhantomData<&'ctx Context>,
+}
+
+impl<'ctx> AsDIScope<'ctx> for DISubProgram<'ctx> {
+    fn as_debug_info_scope(self) -> DIScope<'ctx> {
+        DIScope {
+            metadata_ref: self.metadata_ref,
+            _marker: PhantomData,
+        }
+    }
+}
+
+/// Any kind of debug info type
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub struct DIType<'ctx> {
```

Comment:
In LLVM C++ `DIType`s is derived from `DIScope`. Any reason not to have the DITypes implement `AsDIScope`?